modules/programs/gradle.nix

```diff
@@ -0,0 +1,134 @@
+{ config, pkgs, lib, ... }:
+
+with lib;
+
+let
+  cfg = config.programs.gradle;
+  defaultHomeDirectory = ".gradle";
+  settingsFormat = pkgs.formats.javaProperties { };
+
+  initScript = types.submodule ({ config, ... }: {
+    options = {
+      name = mkOption {
+        type = types.str;
+        description = ''
+          The name of the init script. Has to end with either '.gradle' for
+          Groovy init scripts or '.init.gradle.kts' for Kotlin init scripts.
+          See https://docs.gradle.org/current/userguide/init_scripts.html for
+          more details.
+        '';
+      };
+      text = mkOption {
+        type = types.lines;
+        default =
+          if config.source != null then builtins.readFile config.source else "";
+        defaultText = literalExpression
+          "if source is defined, the content of source, otherwise empty";
+        description = ''
+          Text of the init script file.
+          If unset then the source option will be preferred.
+        '';
+      };
+
+      source = mkOption {
+        type = types.nullOr types.path;
+        default = null;
+        description = ''
+          Path of the init script file to use.
+          If the text option is set, it will be preferred.
+        '';
+      };
+    };
+  });
+in {
+  meta.maintainers = [ maintainers.britter ];
+
+  options.programs.gradle = {
+    enable = mkEnableOption "Gradle Build Tool";
+    home = mkOption {
+      type = types.str;
+      default = defaultHomeDirectory;
+      description = ''
+        The Gradle home directory, relative to `home.homeDirectory`.
+
+        If set, the GRADLE_USER_HOME environment variable will be
+        set accordingly. Defaults to .gradle.
+      '';
+    };
+    package = mkPackageOption pkgs "gradle" { example = "pkgs.gradle_7"; };
+    settings = mkOption {
+      type = types.submodule { freeformType = settingsFormat.type; };
+      default = { };
+      example = {
+        "org.gradle.caching" = true;
+        "org.gradle.parallel" = true;
+        "org.gradle.jvmargs" = "-XX:MaxMetaspaceSize=384m";
+        "org.gradle.home" = pkgs.jdk17;
+      };
+      description = ''
+        Key value pairs to write to gradle.properties in the Gradle home directory.
+      '';
+    };
+    initScripts = mkOption {
+      type = types.either types.path (types.listOf initScript);
```

Comment:
Perhaps `type = with types; attrsOf initScript`?

Would look something like

``` nix
initScripts = {
  "maven-local.gradle".text = "…";
  "another.init.gradle.kts".source = ./another.init.gradle.kts;
};
```

Reply:
To be clear, with the above I also suggest removing the possibility to specify just a path. Instead always require an attribute set. It will help future problems with merging. Specifically, it is unclear what

``` nix
mkMerge [
  { initScripts = ./path; }
  { initScripts."extra-init.gradle".source = ./path2; }
]
```

should do.

You could use `apply` to coerce `./path` to `{ source = ./path; }` but I generally think it is best to avoid `apply` to maintain predictability w.r.t merging.

Reply:
Hey @rycee 👋🏻

My thinking was that if I had multiple init scripts, I'd like to reference a directory instead of listing them all. But I now understand that this doesn't work with merging configurations. I will change this.

Reply:
Done